iroha_python/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+// Allow panic because of bad and unsafe pyo3
+#![allow(clippy::panic)]
+
+use std::ops::{Deref, DerefMut};
+
+use iroha_client::{client, config::Configuration};
+use iroha_crypto::{Hash, PrivateKey, PublicKey};
+use iroha_data_model::prelude::*;
+use pyo3::exceptions::PyException;
+use pyo3::prelude::*;
+
+macro_rules! wrap_class {
+    (
+        $(
+            $ty:ident {
+                $field:ident : $outer_ty:ty
+            } : $( $derive:ident $(+)? )*
+        ),*
+        $(,)?
+    ) => {$(
+        #[pyclass]
+        #[derive($($derive,)*)]
+        pub struct $ty {
+            $field: $outer_ty,
+        }
+
+        impl Deref for $ty {
+            type Target = $outer_ty;
+            fn deref(&self) -> &Self::Target {
+                &self.$field
+            }
+        }
+
+        impl DerefMut for $ty {
+            fn deref_mut(&mut self) -> &mut Self::Target {
+                &mut self.$field
+            }
+        }
+
+        impl From<$outer_ty> for $ty {
+            fn from(outer: $outer_ty) -> Self {
+                Self {
+                    $field: outer,
+                }
+            }
+        }
+
+        impl From<$ty> for $outer_ty {
+            fn from(from: $ty) -> Self {
+                from.$field
+            }
+        }
+
+        #[pymethods]
+        impl $ty {
+            fn __str__(&self) -> String {
+                format!("{:#?}", self)
+            }
+        }
+    )*
+        fn register_wrapped_classes(m: &PyModule) -> PyResult<()> {
+            $(m.add_class::<$ty>()?;)*
+            Ok(())
+        }
+    };
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct Dict<T>(T);
+
+impl<T> Deref for Dict<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> DerefMut for Dict<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<T> Dict<T> {
+    #[allow(clippy::missing_const_for_fn)]
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<'source, T: serde::de::DeserializeOwned> FromPyObject<'source> for Dict<T> {
+    fn extract(obj: &'source PyAny) -> PyResult<Self> {
+        let obj = if obj.hasattr("to_rust")? {
+            obj.call_method0("to_rust")?
```

Comment:
It would be good to have "to_rust" str as a constant, it is referenced in several places.

Reply:
It is used in single place here. You won't be able to synchronize this contract with python anyway, so i think there is no need for constant.

Reply:
I mean it will anyway be soft contract and it is referenced only in 2 lines which are tighed up together.

Reply:
Yes, I see, the other one was in python

Reply:
Ok